Code/GraphMol/MolOps.cpp

```diff
@@ -1168,5 +1168,105 @@ std::vector<int> hapticBondEndpoints(const Bond *bond) {
   return oats;
 }
 }  // end of namespace details
-};  // end of namespace MolOps
-};  // end of namespace RDKit
+
+void expandAttachmentPoints(RWMol &mol, bool addAsQueries) {
+  for (auto atom : mol.atoms()) {
+    int value;
+    if (atom->getPropIfPresent(common_properties::molAttachPoint, value)) {
+      std::vector<int> tgtVals;
+      if (value == 1 || value == -1) {
+        tgtVals.push_back(1);
+      }
+      if (value == 2 || value == -1) {
+        tgtVals.push_back(2);
+      }
+      if (tgtVals.empty()) {
+        BOOST_LOG(rdWarningLog)
+            << "Invalid value for molAttachPoint: " << value << " on atom "
+            << atom->getIdx() << ". Not expanding this atttachment point."
+            << std::endl;
+        continue;
+      }
+      for (auto tval : tgtVals) {
+        atom->clearProp(common_properties::molAttachPoint);
+        Atom *newAtom = nullptr;
+        if (addAsQueries) {
+          newAtom = new QueryAtom(0);
+          newAtom->setQuery(RDKit::makeAtomNullQuery());
+        } else {
+          newAtom = new Atom(0);
+        }
+        newAtom->setProp(common_properties::_fromAttachPoint, tval);
+        bool updateLabel = false;
+        bool takeOwnership = true;
+        auto idx = mol.addAtom(newAtom, updateLabel, takeOwnership);
+        mol.addBond(atom->getIdx(), idx, Bond::SINGLE);
+      }
+    }
+  }
+}
+
+void collapseAttachmentPoints(RWMol &mol, bool markedOnly) {
+  bool removedAny = false;
+  std::vector<int> attachLabels(mol.getNumAtoms(), 0);
+  for (auto atom : mol.atoms()) {
+    if (atom->getAtomicNum() == 0) {
+      int value = 0;
+      atom->getPropIfPresent(common_properties::_fromAttachPoint, value);
```

Comment:
If I'm reading this right, reading in `*OC |$_AP1;;$|` wouldn't assign _fromAttachPoint to the dummy and this wouldn't collapse. Shouldn't it though?

And would it be possible to add a public API that does this check? 
```
bool is_attachment_point_atom(const RDKit::Atom *const atom);
```
Our logic is:
- is a dummy atom (atno == 0)
- total degree == 1
- atomLabel is set to "_AP%d"

but I can see how that would be expanded to the additional logic you have below

Reply:
I need to look at this one in more detail